tests/utilities/test_auto_restart.py

```diff
@@ -0,0 +1,640 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+import os
+import random
+from collections.abc import Iterable
+from typing import Optional
+
+import numpy as np
+import pytest
+import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+from torch.utils.data import BatchSampler, DistributedSampler, RandomSampler, SequentialSampler
+from torch.utils.data._utils.worker import get_worker_info
+from torch.utils.data.dataloader import DataLoader, default_collate
+from torch.utils.data.dataset import Dataset, IterableDataset
+
+import tests.helpers.utils as tutils
+from pytorch_lightning import seed_everything
+from pytorch_lightning.utilities.auto_restart import CaptureIterativeDataset, FastForwardSampler
+from pytorch_lightning.utilities.data import has_len
+from pytorch_lightning.utilities.enums import AutoRestartBatchKeys
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from tests.helpers.runif import RunIf
+
+
+# Credit to PyTorch Team.
+# Taken from:
+# https://github.com/pytorch/pytorch/blob/3b977a0d2834d300c0301a0c6af98c8e939019ce/torch/utils/data/_utils/worker.py#L151
+# Not available in PyTorch 1.4.
+def _generate_state(base_seed, worker_id):
```

Comment:
since we will drop 1.4 in a few days, it would probably make more sense to directly import from pytorch, right?

Reply:
I removed it :)